DeviceAdapters/IntegratedLaserEngine/PortsConfiguration.cpp

```diff
@@ -0,0 +1,261 @@
+///////////////////////////////////////////////////////////////////////////////
+// FILE:          PortsConfiguration.cpp
+// PROJECT:       Micro-Manager
+// SUBSYSTEM:     DeviceAdapters
+//-----------------------------------------------------------------------------
+
+#include "PortsConfiguration.h"
+#include "IntegratedLaserEngine.h"
+#include <windows.h>
+#include <Shlobj.h>
+#include <fstream>
+#include "boost\filesystem.hpp"
+
+
+#if defined(WIN32) || defined(_WIN32) 
+#define PATH_SEPARATOR "\\" 
+#else 
+#define PATH_SEPARATOR "/" 
+#endif 
+
+const std::string CPortsConfiguration::ConfigFileKeyBaseName_ = "OutputPort";
+
+
+CPortsConfiguration::CPortsConfiguration( const std::string& Device1Name, const std::string& Device2Name, CIntegratedLaserEngine* MMILE ) :
+  Device1Name_( Device1Name ),
+  Device2Name_( Device2Name ),
+  MMILE_( MMILE )
+{  
+  if ( MMILE_ == nullptr )
+  {
+    throw std::logic_error( "CPortsConfiguration: Pointer tomain class invalid" );
+  }
+
+  bool vGenerateDefaultConfig = false;
+  std::string vFile1Name = GenerateFileName( Device1Name_, Device2Name_ );
+  std::string vFile2Name = GenerateFileName( Device2Name_, Device1Name_ );
+  // Look for the config file in the MicroManager folder
+  if ( !LoadConfigFile( vFile1Name ) )
+  {
+    if ( !LoadConfigFile( vFile2Name ) )
+    {
+      // Look for the config file in the Fusion config folder
+      std::string vFusionConfigFolderPath;
+      if ( GetFusionConfigFolderPath( &vFusionConfigFolderPath ) )
+      {
+        std::string vFileName = vFusionConfigFolderPath + vFile1Name;
+        if ( !LoadConfigFile( vFileName ) )
+        {
+          vFileName = vFusionConfigFolderPath + vFile2Name;
+          if ( !LoadConfigFile( vFileName ) )
+          {
+            // No config file found, generate a demo file
+            vGenerateDefaultConfig = true;
+            MMILE_->LogMMMessage( "No port configuration file found. Using a default port configuration." );
+          }
+        }
+      }
+      else
+      {
+        vGenerateDefaultConfig = true;
+      }
+    }
+  }
+  if ( vGenerateDefaultConfig )
+  {
+    GenerateDefaultConfig();
+  }
+}
+
+CPortsConfiguration::~CPortsConfiguration()
+{
+}
+
+std::string CPortsConfiguration::GenerateFileName( const std::string& Device1Name, const std::string& Device2Name ) const
+{
+  static const std::string vBaseFileName = "DualILE - ";
+  return vBaseFileName + Device1Name + ", " + Device2Name + ".desc";
+}
+
+bool CPortsConfiguration::GetFusionConfigFolderPath( std::string* Path ) const
+{
+  PWSTR vWPathName;
+  if ( SUCCEEDED( SHGetKnownFolderPath( FOLDERID_PublicDocuments, KF_FLAG_DEFAULT, NULL, &vWPathName ) ) ) // FOLDERID_PublicDocuments for Public documents path - FOLDERID_ProgramData for All users (aka ProgramData) path
+  {
+    std::wstring vWPathNameString = vWPathName;
+    std::string vBaseFileName( vWPathNameString.begin(), vWPathNameString.end() );
```

Comment:
Here I get the warning

```text
C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\xstring(2541,23): warning C4244: 'argument': conversion from 'wchar_t' to 'const _Elem', possible loss of data
        with
        [
            _Elem=char
        ]
```

as a result of passing a pair of `wchar_t` iterators to the constructor of `std::string` (it will not autoconvert but rather take the low byte of each wide character).

I think this is safe if the string is guaranteed to be ASCII only. Is that a good assumption here (e.g. on non-English versions of Windows)?

Reply:
I improved this the same way as in the Dragonfly adapter using ```wcstombs_s```.